File: src/parser/typeParser.ts
import type { TypeExpr } from "../ast.js";
import { tokenize } from "./lexer.js";
import type { Token } from "./lexer.js";

type ParseState = { tokens: Token[]; pos: number };

export function parseTypeExpr(source: string, file: string): TypeExpr {
  const tokens = tokenize(source, file);
  // Remove EOF token for cleaner parsing
  const state: ParseState = { tokens: tokens.slice(0, -1), pos: 0 };
  return parseSumType(state);
}

function peek(s: ParseState): Token {
  return s.tokens[s.pos] ?? { kind: "eof", value: "", span: { file: "", startLine: 1, startColumn: 1, endLine: 1, endColumn: 1 } };
}

function advance(s: ParseState): Token {
  return s.tokens[s.pos++]!;
}

// Sum type: Type | Type | ...
function parseSumType(s: ParseState): TypeExpr {
  let left = parseFunctionType(s);
  while (peek(s).kind === "pipe") {
    advance(s); // |
    const right = parseFunctionType(s);
    if (left.kind === "sumType") {
      left.variants.push(right);
    } else {
      left = { kind: "sumType", variants: [left, right], span: mergeTypeSpan(left, right) };
    }
  }
  return left;
}

// Function type: (Params) => Return  OR  Type => Return
function parseFunctionType(s: ParseState): TypeExpr {
  if (peek(s).kind === "lparen") {
    advance(s); // (
    const params: TypeExpr[] = [];
    while (peek(s).kind !== "rparen" && peek(s).kind !== "eof") {
      params.push(parseSumType(s));
      if (peek(s).kind === "comma") advance(s);
    }
    if (peek(s).kind === "rparen") advance(s); // )

    if (peek(s).kind === "fatArrow") {
      advance(s); // ->
      const returns = parseSumType(s);
      const span = params.length > 0
        ? { ...params[0]!.span, endLine: returns.span.endLine, endColumn: returns.span.endColumn }
        : returns.span;
      return { kind: "functionType", params, returns, span };
    }
    // Just parenthesized type
    if (params.length === 1) return params[0]!;
    return { kind: "namedType", name: "void", span: { file: "", startLine: 1, startColumn: 1, endLine: 1, endColumn: 1 } };
  }

  let left = parseGenericType(s);
  if (peek(s).kind === "fatArrow") {
    advance(s);
    const returns = parseSumType(s);
    return { kind: "functionType", params: [left], returns, span: mergeTypeSpan(left, returns) };
  }
  return left;
}

// Generic: Name<Args>
function parseGenericType(s: ParseState): TypeExpr {
  const left = parsePrimaryType(s);
  if (peek(s).kind === "lt") {
    advance(s); // <
    const args: TypeExpr[] = [];
    while (peek(s).kind !== "gt" && peek(s).kind !== "eof") {
      args.push(parseSumType(s));
      if (peek(s).kind === "comma") advance(s);
    }
    if (peek(s).kind === "gt") advance(s); // >
    const baseName = left.kind === "namedType" ? left.name : left.kind === "primitiveType" ? left.name : "";
    return { kind: "genericType", base: baseName, typeArgs: args, span: left.span };
  }
  return left;
}

// Primary: String | Int | Bool | Identifier
function parsePrimaryType(s: ParseState): TypeExpr {
  const token = advance(s);
  if (token.kind === "identifier") {
    if (["String", "Int", "Bool"].includes(token.value)) {
      return { kind: "primitiveType", name: token.value as "String" | "Int" | "Bool", span: token.span };
    }
    return { kind: "namedType", name: token.value, span: token.span };
  }
  return { kind: "namedType", name: "unknown", span: token.span };
}

function mergeTypeSpan(a: TypeExpr, b: TypeExpr): TypeExpr["span"] {
  return {
    file: a.span.file,
    startLine: a.span.startLine,
    startColumn: a.span.startColumn,
    endLine: b.span.endLine,
    endColumn: b.span.endColumn
  };
}

File: compiler/tangle-cli/src/parser/type_parser.rs
use crate::ast::*;
use crate::model::{SourceSpan, TangleDiagnostic};
use crate::parser::lexer::{tokenize, Token, TokenKind};

// ============================================================
// Public entry point
// ============================================================

pub fn parse_type_expr(source: &str, file: &str) -> (Option<TypeExpr>, Vec<TangleDiagnostic>) {
    let (tokens, lexer_diags) = tokenize(source, file);
    let mut parser = TypeParser::new(&tokens);
    let result = parser.parse_sum_type();
    let mut diags = lexer_diags;
    diags.extend(parser.diagnostics);
    (result, diags)
}

// ============================================================
// TypeParser
// ============================================================

struct TypeParser<'a> {
    tokens: &'a [Token],
    pos: usize,
    diagnostics: Vec<TangleDiagnostic>,
}

impl<'a> TypeParser<'a> {
    fn new(tokens: &'a [Token]) -> Self {
        TypeParser {
            tokens,
            pos: 0,
            diagnostics: vec![],
        }
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn advance(&mut self) -> Token {
        let t = self.tokens[self.pos].clone();
        if self.pos + 1 < self.tokens.len() {
            self.pos += 1;
        }
        t
    }

    // ============================================================
    // Sum type: T1 | T2 | T3
    // ============================================================

    fn parse_sum_type(&mut self) -> Option<TypeExpr> {
        let mut left = self.parse_function_type()?;
        while self.peek().kind == TokenKind::Pipe {
            self.advance(); // consume '|'
            let right = self.parse_function_type()?;
            left = match left {
                TypeExpr::Sum(mut s) => {
                    let right_span = span_of_type(&right);
                    s.variants.push(right);
                    s.span = merge_spans(&s.span, &right_span);
                    TypeExpr::Sum(s)
                }
                other => {
                    let right_span = span_of_type(&right);
                    let span = merge_spans(&span_of_type(&other), &right_span);
                    TypeExpr::Sum(SumTypeExpr {
                        variants: vec![other, right],
                        span,
                    })
                }
            };
        }
        Some(left)
    }

    // ============================================================
    // Function type: Param -> Return
    // ============================================================

    fn parse_function_type(&mut self) -> Option<TypeExpr> {
        let left = self.parse_generic_or_primary()?;
        if self.peek().kind == TokenKind::Arrow {
            self.advance(); // consume '->'
            let returns = self.parse_sum_type()?;
            let params = match &left {
                TypeExpr::Primitive(_) | TypeExpr::Named(_) | TypeExpr::Generic(_) => {
                    vec![left.clone()]
                }
                TypeExpr::Sum(inner) => {
                    // For sum types in param position, each variant is treated as a parameter
                    // e.g., String | Int -> Bool  =>  (String | Int) -> Bool
                    vec![TypeExpr::Sum(inner.clone())]
                }
                _ => vec![left.clone()],
            };
            let span = merge_spans(&span_of_type(&params[0]), &span_of_type(&returns));
            return Some(TypeExpr::Function(FunctionTypeExpr {
                params,
                returns: Box::new(returns),
                span,
            }));
        }
        Some(left)
    }

    // ============================================================
    // Generic type: Result<String, Error>  or  Array<Int>
    // ============================================================

    fn parse_generic_or_primary(&mut self) -> Option<TypeExpr> {
        let name_tok = self.advance();
        if name_tok.kind != TokenKind::Identifier {
            self.diagnostics.push(TangleDiagnostic {
                code: "TANGLE_TYPE_PARSE_ERROR".into(),
                message: format!(
                    "Expected type name, got: {:?}",
                    name_tok.kind
                ),
                span: name_tok.span.clone(),
            });
            return None;
        }
        let name = name_tok.lexeme.clone();
        let span = name_tok.span.clone();

        // Check for generic type args: Name<A, B, ...>
        if self.peek().kind == TokenKind::Lt {
            self.advance(); // consume '<'
            let mut args = vec![];
            loop {
                args.push(self.parse_sum_type()?);
                if self.peek().kind == TokenKind::Comma {
                    self.advance(); // consume ','
                    continue;
                }
                if self.peek().kind == TokenKind::Gt {
                    self.advance(); // consume '>'
                    break;
                }
                self.diagnostics.push(TangleDiagnostic {
                    code: "TANGLE_TYPE_PARSE_ERROR".into(),
                    message: "Expected ',' or '>' in type arguments".into(),
                    span: self.peek().span.clone(),
                });
                return None;
            }
            let end_span = self.tokens[self.pos - 1].span.clone();
            return Some(TypeExpr::Generic(GenericTypeExpr {
                base: name,
                type_args: args,
                span: merge_spans(&span, &end_span),
            }));
        }

        // Distinguish primitive types from named (user-defined) types
        match name.as_str() {
            "String" | "Int" | "Float" | "Bool" => {
                Some(TypeExpr::Primitive(PrimitiveTypeExpr { name, span }))
            }
            _ => Some(TypeExpr::Named(NamedTypeExpr { name, span })),
        }
    }
}

// ============================================================
// Helpers
// ============================================================

fn span_of_type(ty: &TypeExpr) -> SourceSpan {
    match ty {
        TypeExpr::Primitive(t) => t.span.clone(),
        TypeExpr::Sum(t) => t.span.clone(),
        TypeExpr::Generic(t) => t.span.clone(),
        TypeExpr::Function(t) => t.span.clone(),
        TypeExpr::Named(t) => t.span.clone(),
    }
}

fn merge_spans(start: &SourceSpan, end: &SourceSpan) -> SourceSpan {
    SourceSpan {
        file: start.file.clone(),
        start_line: start.start_line,
        start_column: start.start_column,
        end_line: end.end_line,
        end_column: end.end_column,
    }
}
